### ISSUE TITLE
sfdisk --wipe=always does not wipe old disk label

### ISSUE THREAD

**Issue by marcosfrm** [karelzak/util-linux#410]
```
# wipefs /dev/sdb    # shows nothing

# echo ',,L' | sfdisk --wipe=always --label=gpt /dev/sdb
Checking that no-one is using this disk right now ... OK

Disk /dev/sdb: 232.9 GiB, 250059350016 bytes, 488397168 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes

>>> Created a new GPT disklabel (GUID: 39B32E92-E427-4380-98FD-3BAB34A32442).
/dev/sdb1: Created a new partition 1 of type 'Linux filesystem' and of size 232.9 GiB.
/dev/sdb2: Done.

New situation:
Disklabel type: gpt
Disk identifier: 39B32E92-E427-4380-98FD-3BAB34A32442

Device     Start       End   Sectors   Size Type
/dev/sdb1   2048 488397134 488395087 232.9G Linux filesystem

The partition table has been altered.
Calling ioctl() to re-read partition table.
Syncing disks.

# wipefs /dev/sdb
offset               type
----------------------------------------------------------------
0x200                gpt   [partition table]
```

Now, create a DOS disk label:

```
# echo ',,L' | sfdisk --wipe=always --label=dos /dev/sdb
Checking that no-one is using this disk right now ... OK

Disk /dev/sdb: 232.9 GiB, 250059350016 bytes, 488397168 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: 39B32E92-E427-4380-98FD-3BAB34A32442

Old situation:

Device     Start       End   Sectors   Size Type
/dev/sdb1   2048 488397134 488395087 232.9G Linux filesystem

>>> Created a new DOS disklabel with disk identifier 0x0bfa5e25.
/dev/sdb1: Created a new partition 1 of type 'Linux' and of size 232.9 GiB.
/dev/sdb2: Done.

New situation:
Disklabel type: dos
Disk identifier: 0x0bfa5e25

Device     Boot Start       End   Sectors   Size Id Type
/dev/sdb1        2048 488397167 488395120 232.9G 83 Linux

The partition table has been altered.
Calling ioctl() to re-read partition table.
Syncing disks.
```

But...

```
# wipefs /dev/sdb
offset               type
----------------------------------------------------------------
0x200                gpt   [partition table]

0x1fe                dos   [partition table]
```

Should `--wipe=always` nuke the gpt label?

util-linux 2.29.1 here.

**Comment by marcosfrm** [comment]
According to the man page it should:

> -w, --wipe when
> Wipe filesystem, RAID and partition-table signatures from the device, in order to avoid possible collisions.

**Comment by marcosfrm** [comment]
It is really broken:

```
# wipefs /dev/sdb    # shows nothing

# mkfs.jfs -q /dev/sdb                                                                                                                   :(
mkfs.jfs version 1.1.15, 04-Mar-2011
   |

Format completed successfully.

244198584 kilobytes total disk space.

# echo ',,L' | sfdisk --wipe=always --label=dos /dev/sdb
Checking that no-one is using this disk right now ... OK

Device /dev/sdb already contains a jfs signature.
The signature will be removed by a write command.

Disk /dev/sdb: 232.9 GiB, 250059350016 bytes, 488397168 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes

>>> Created a new DOS disklabel with disk identifier 0x57f7f22b.
/dev/sdb1: Created a new partition 1 of type 'Linux' and of size 232.9 GiB.
/dev/sdb2: Done.

New situation:
Disklabel type: dos
Disk identifier: 0x57f7f22b

Device     Boot Start       End   Sectors   Size Id Type
/dev/sdb1        2048 488397167 488395120 232.9G 83 Linux

The partition table has been altered.
Calling ioctl() to re-read partition table.
Syncing disks.

# wipefs /dev/sdb
offset               type
----------------------------------------------------------------
0x1fe                PMBR   [partition table]

0x8000               jfs   [filesystem]
                     UUID:  fada279b-acf2-403a-bb2f-c1219b85d303
```

**Comment by karelzak** [comment]
Hmm... the description in the man page is pretty bad, it does not match how libfdisk uses collisions detection. The detection is enabled only if the disk is not partitioned yet, because already partitioned disk is standard use-case (for example if you want to modify/add a partition only).

I agree that for "always" it's strange semantic and wipe should not depend on the default collisions detection in the library. Going to fix it...

The example with mkfs.jfs is obvious bug.

**Comment by karelzak** [comment]
Note that the patch only fixes the issue with mkfs only, not the issue with PT collisions.

**Comment by karelzak** [comment]
All the issues should be fixed now. I have added regression tests to be sure we will not see the issue again.

Thanks!

**Comment by marcosfrm** [comment]
Thanks  @karelzak  ! Can we have these fixes in 2.29.2?

**Comment by karelzak** [comment]
Not sure if all the patches, but I'll try to use as much as possible.
